exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -448,39 +427,44 @@ public Builder withValidators(short validatorCount) {
     }
 
     /**
-     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     * Adds a service artifact which would be deployed by the TestKit. Several service artifacts
+     * can be added.
+     *
+     * <p>Note that the corresponding service instance with equal serviceArtifactId should be
+     * created with {@link #withService(ServiceArtifactId, String, int, Any)}.
      */
-    public Builder withService(Class<? extends ServiceModule> serviceModule) {
-      services.add(serviceModule);
+    public Builder withDeployedService(ServiceArtifactId serviceArtifactId, String artifactFilename) {
+      serviceArtifactFilenames.put(serviceArtifactId, artifactFilename);
       return this;
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
-     */
-    @SafeVarargs
-    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
-                                      Class<? extends ServiceModule>... serviceModules) {
-      return withServices(Lists.asList(serviceModule, serviceModules));
-    }
-
-    /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance. Several service specifications can be added.
```

Comment:
All services are started and configured before the genesis block
(please check if that's accurate. Also, what would happen if configuration of such service fails? Will I get an exception or just silent start failure, as with 'dynamic' services? I'd test that)

Reply:
Yes - TestKit services are created as builtin services, so they are instantiated before genesis block is committed - https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/blockchain/builder.rs#L140